# Исследование надежности заемщиков


Во второй части проекта вы выполните шаги 3 и 4. Их вручную проверит ревьюер.
Чтобы вам не пришлось писать код заново для шагов 1 и 2, мы добавили авторские решения в ячейки с кодом. 



## Откройте таблицу и изучите общую информацию о данных

**Задание 1. Импортируйте библиотеку pandas. Считайте данные из csv-файла в датафрейм и сохраните в переменную `data`. Путь к файлу:**

`/datasets/data.csv`

In [1]:
import pandas as pd

try:
    data = pd.read_csv('/datasets/data.csv')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/data.csv')

**Задание 2. Выведите первые 20 строчек датафрейма `data` на экран.**

In [2]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

**Задание 3. Выведите основную информацию о датафрейме с помощью метода `info()`.**

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Предобработка данных

### Удаление пропусков

**Задание 4. Выведите количество пропущенных значений для каждого столбца. Используйте комбинацию двух методов.**

In [4]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**Задание 5. В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце вы обработаете на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [5]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

**Задание 6. В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Для реальных данных это нормально. Обработайте значения в этом столбце: замените все отрицательные значения положительными с помощью метода `abs()`.**

In [6]:
data['days_employed'] = data['days_employed'].abs()

**Задание 7. Для каждого типа занятости выведите медианное значение трудового стажа `days_employed` в днях.**

In [7]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьте их как есть. Тем более этот столбец не понадобится вам для исследования.

**Задание 8. Выведите перечень уникальных значений столбца `children`.**

In [8]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

**Задание 9. В столбце `children` есть два аномальных значения. Удалите строки, в которых встречаются такие аномальные значения из датафрейма `data`.**

In [9]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

**Задание 10. Ещё раз выведите перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [10]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5])

### Удаление пропусков (продолжение)

**Задание 11. Заполните пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**

In [11]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

**Задание 12. Убедитесь, что все пропуски заполнены. Проверьте себя и ещё раз выведите количество пропущенных значений для каждого столбца с помощью двух методов.**

In [12]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

**Задание 13. Замените вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [13]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

**Задание 14. Обработайте неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведите их к нижнему регистру. Проверьте остальные столбцы.**

In [14]:
data['education'] = data['education'].str.lower()
data['family_status'] = data['family_status'].str.lower()

**Задание 15. Выведите на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалите их.**

In [15]:
data.duplicated().sum()

71

In [16]:
data = data.drop_duplicates()

### Категоризация данных

**Задание 16. На основании диапазонов, указанных ниже, создайте в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используйте собственную функцию с именем `categorize_income()` и метод `apply()`.**

In [17]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [18]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

**Задание 17. Выведите на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [19]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Задание 18. Создайте функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.**

**Используйте собственную функцию с именем `categorize_purpose()` и метод `apply()`. Изучите данные в столбце `purpose` и определите, какие подстроки помогут вам правильно определить категорию.**

In [20]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [21]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

**Создание функции, которая на основании данных из столбца days_employed сформирует новый столбец days_employed_category, в который войдут следующие категории: A, B, C, D, E, Unreal**

In [22]:
def categorize_days_employed(days_employed):
    try:
        if 0 <= days_employed <= 500:
            return 'E'
        elif 501 <= days_employed <= 1500:
            return 'D'
        elif 1501 <= days_employed <= 3000:
            return 'C'
        elif 3001 <= days_employed <= 5000:
            return 'B'
        elif 5001 <= days_employed <= 10000:
            return 'A'
        else:
            return 'Unreal'
    except:
        pass

In [23]:
data['days_employed_category'] = data['days_employed'].apply(categorize_days_employed)

**Создание функции, которая на основании данных из столбца dob_years сформирует новый столбец dob_years_category, в который войдут следующие категории: A, B, C, D, E, Unreal**

In [24]:
def categorize_dob_years(dob_years):
    try:
        if 0 <= dob_years <= 18:
            return 'A'
        elif 19 <= dob_years <= 27:
            return 'B'
        elif 28 <= dob_years <= 50:
            return 'C'
        elif 51 <= dob_years <= 75:
            return 'D'
        elif 76 <= dob_years:
            return 'E'
        else:
            return 'Unreal'
    except:
        pass

In [25]:
data['dob_years_category'] = data['dob_years'].apply(categorize_dob_years)

### Шаг 3. Исследуйте данные и ответьте на вопросы

**Создание функции, которая рассчитывает вероятность задолженности для каждого уникального значения в столбце column_to_check.**
    Аргументы:
        column_to_check: название анализируемого столбца.
    Возвращает:
        DataFrame с данными.

In [26]:
def probability_of_debt_old(col1, col2):
    col1_name = col1.name
    col2_name = col2.name
    in_data = pd.DataFrame({col1_name: col1, col2_name: col2})

    for t in sorted(in_data[col1_name].unique()):
        debts = in_data.loc[in_data[col1_name]==t, col2_name].sum()
        total = in_data.loc[in_data[col1_name]==t, col2_name].count()
        prob_of_debt = debts/total
        
        if total != 0:
            prob_of_debt = debts / total
            print(f"Вероятность задолженности для {col1_name} = {t} составляет: {prob_of_debt*100:.2f} %. \n(всего {total} займа - {total/len(col1)*100:.2f} % выборки)")
        else:
            print(f"Для {col1_name} = {t} отсутствуют сведения.")


In [27]:
def probability_of_debt_ver1(column_to_check):
    
    probability = (data.groupby([column_to_check])['debt'].mean()*100).sort_values(ascending=False).to_frame (name='debt probability (%)')
    quantity = (data.groupby([column_to_check])['debt'].count()).to_frame (name='quantity')
    relative_quantity = (data.groupby([column_to_check])['debt'].count() / len(data)*100).to_frame (name='relative quantity (%)')

    prob_df = probability.merge(quantity, on=column_to_check, how='left')
    prob_df = prob_df.merge(relative_quantity, on=column_to_check, how='left')
    
    return round(prob_df, 2)

In [28]:
def probability_of_debt_ver2(column_to_check):
    
    df = data.groupby([column_to_check]).agg({'debt' : ['mean', 'count', relative_quantity]})
    
    df.columns = ['debt probability (%)', 'quantity', 'relative quantity (%)']
    df['debt probability (%)'] = round(df['debt probability (%)']*100, 2)
    df = df.sort_values('debt probability (%)', ascending=False)
    return df

def relative_quantity(q):
    relative_quantity = round(100*len(q) / len(data), 2)
    return relative_quantity

In [29]:
def probability_of_debt_ver3(column_to_check):
    data_pivot = data.pivot_table(index=column_to_check, values='debt', aggfunc=['mean', 'count', relative_quantity])
    data_pivot['mean'] = round(100*data_pivot['mean'], 2)
    data_pivot.columns = ['debt probability (%)', 'quantity', 'relative quantity (%)']
    data_pivot = data_pivot.sort_values('debt probability (%)', ascending=False)
    return data_pivot

In [30]:
#%%timeit
#probability_of_debt_old(data['children'], data['debt'])
#probability_of_debt_ver1('children')
#probability_of_debt_ver2('children')
#probability_of_debt_ver3('children')

Наиболее эффективная функция, возвращаюющая df: probability_of_debt_ver2 (это groupby() совиместно с agg()).\
Наименее эффективная функция, возвращаюющая df: probability_of_debt_ver3 (это pivot_table()).

<div class="alert alert-danger">
<font size="5"><b>Комментарий ревьюера</b></font>

Ошибка: Функция правильная, но такой способ реализации всех расчетов крайне неэффективный. Также не стоит выводить данные построчно. Лучше сразу привыкать выводить информацию в виде таблиц. Именно так и ведется работа с данными. Для применения методов по уникальным параметрам столбца следует использовать метод groupby. Так мы проводим группировку по значениям выбранного параметра с возможностью расчетов по ним. Попробуй использовать именно его для получения результатов. 

</div>

<div class="alert alert-success">
<font size="5"><b>Комментарий ревьюера 2</b></font>

Успех: Отличная реализация получилась! 

</div>

#### 3.1 Есть ли зависимость между количеством детей и возвратом кредита в срок?

In [31]:
#data.groupby(['children'])['debt'].mean()

In [32]:
probability_of_debt_ver1('children')

debt probability (%)  quantity  relative quantity (%)
children                                                       
4                         9.76        41                   0.19
2                         9.45      2052                   9.62
1                         9.23      4808                  22.54
3                         8.18       330                   1.55
0                         7.54     14091                  66.06
5                         0.00         9                   0.04

**Вывод:**

Согласно проведенным рассчетам, между количеством детей и возвратом кредита в срок существует зависмость.

В зависмости от количества детей вероятность задолженности составляет от 7.54 % до 9.76 %. При этом не целесообразно учитывать вероятность 0 % для категории с 5 детьми ввиду малого количества сведений по указанной категории в представленной выборке (всего 9 займов).

Наименьшая вероятность задолженности у категории заемщиков без детей (7.54 %). При этом, заемщики указанной категории составляют большую часть представленной выборки (66.06 %).

Наибольшая вероятность отмечается в категории с 4 детьми (9.76 %).

При этом сделать вывод о том, что вероятность задолженности увеличивается с количеством детей не представляется возможным.

<div class="alert alert-success">
<font size="5"><b>Комментарий ревьюера</b></font>

Успех: Вывод не противоречит полученному результату. С ним все хорошо, а вот сам метод расчета следует скорректировать. Весь вывод информации в идеале вполне можно уложить в одну строку кода. 

</div>

#### 3.2 Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [33]:
#data.groupby(['family_status']).agg({'debt' : ['mean', 'count', relative_quantity]})

In [34]:
probability_of_debt_ver2('family_status')

debt probability (%)  quantity  relative quantity (%)
family_status                                                               
не женат / не замужем                  9.76      2796                  13.11
гражданский брак                       9.31      4134                  19.38
женат / замужем                        7.56     12261                  57.48
в разводе                              7.06      1189                   5.57
вдовец / вдова                         6.62       951                   4.46

<div class="alert alert-success">
<font size="5"><b>Комментарий ревьюера</b></font>
  
Совет: Будет здорово, если ты освоишь метод [agg](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.agg.html), который можно применить после группировки и передать в него лист с аггрегирующими функциями: sum - для подсчета числа должников, count - для подсчета общего числа клиентов в группах и mean - для нахождения доли должников в каждой из групп. 

</div>

**Вывод:** 

Согласно проведенным рассчетам, между семейным положением и возвратом кредита в срок существует зависмость.

В зависмости от семейного положения вероятность задолженности составляет от 6.62 % (вдовец / вдова) до 9.76 % (Не женат / не замужем).

При этом наиболее многочисленна категория "женат / замужем" (57.48 % выборки) с верочятностью задолженности 7.56 %.

Кроме того отмечается, что наименования категорий сатуса семейного положения могут вводить в заблуждение. Например, "в разводе" и "вдовец/вдова" могут принадлежать к категории "Не женат / не замужем". И наоборот. Что может повлиять на результаты исследования.

<div class="alert alert-success">
<font size="5"><b>Комментарий ревьюера</b></font>

Успех: Подумай также и над причинами получения такого результата. Код прошу исправить как минимум в одном из пунктов. 

</div>

#### 3.3 Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [35]:
probability_of_debt_ver2('total_income_category')

debt probability (%)  quantity  relative quantity (%)
total_income_category                                                       
E                                      9.09        22                   0.10
C                                      8.50     15921                  74.64
A                                      8.00        25                   0.12
B                                      7.06      5014                  23.51
D                                      6.02       349                   1.64

**Вывод:** 

Согласно проведенным рассчетам, между уровнем дохода и возвратом кредита в срок существует зависмость.

В зависмости от уровня дохода вероятность задолженности составляет от 6.02 % (кат. "D") до 8.50 % (кат."С"). При этом не целесообразно учитывать вероятность для категорий дохода "А" (8.00 %) и "Е" (9.09 %) ввиду малого количества сведений по указанным категориям в представленной выборке (всего 25 и 22 займа соответственно).

Наибольшая вероятность отмечается у заемщиков с категорией дохода "С" (8.50 %). При этом, заемщики указанной категории составляют большую часть представленной выборки (74.64 %).

Сделать вывод о том, что вероятность задолженности уменьшается с увеличением уровня дохода не представляется возможным.

<div class="alert alert-success">
<font size="5"><b>Комментарий ревьюера</b></font>

Успех: Соглашусь, не все группы имеют достаточное число клиентов для построения по ним надежных выводов. 

</div>

#### 3.4 Как разные цели кредита влияют на его возврат в срок?

In [36]:
#data_pivot = data.pivot_table(index='purpose_category', values='debt', aggfunc=['mean', 'count', relative_quantity])
#data_pivot

In [37]:
probability_of_debt_ver3('purpose_category')

debt probability (%)  quantity  \
purpose_category                                           
операции с автомобилем                    9.35      4279   
получение образования                     9.25      3988   
проведение свадьбы                        7.91      2313   
операции с недвижимостью                  7.26     10751   

                          relative quantity (%)  
purpose_category                                 
операции с автомобилем                    20.06  
получение образования                     18.70  
проведение свадьбы                        10.84  
операции с недвижимостью                  50.40

**Вывод:** 

Согласно проведенным рассчетам, между целью кредита и возвратом кредита в срок существует зависмость.

В зависмости от цели кредита вероятность задолженности составляет от 7.26 % (операции с недвижимостью) до 9.35 % (операции с автомобилем).

При этом наиболее многочисленна категория "операции с недвижимостью" (50.40 % выборки).

<div class="alert alert-danger">
<font size="5"><b>Комментарий ревьюера</b></font>

Ошибка: Также на данном шаге рекомендуется использовать метод сводных таблиц (pivot_table()). Однако он является аналогом метода groupby(). Изучи метод сводных таблиц и примени его хотя бы в одном пункте анализа.

</div>

############################################################ \
\
***Артем, я не совсем понял суть применения сводной таблицы в конкретной данной ситуации. Цель была продемонстрировать метод? Или я не так понял комментарий и что-то не то сделал?*** \
\
Пойдёт такой способ "выделить" мои коментарии?\
\
############################################################

<div class="alert alert-success">
<font size="5"><b>Комментарий ревьюера 2</b></font>

Успех: Да, так вполне можно выделить свои комментарии. Суть была именно о использовании метода, все верно. Сводные таблицы нам очень пригодятся в дальнейших проектах.
</div>

#### 3.5 Есть ли зависимость между трудовым стажем и возвратом кредита в срок

In [38]:
probability_of_debt_ver2('days_employed_category')

debt probability (%)  quantity  relative quantity (%)
days_employed_category                                                       
E                                      11.24      2599                  12.18
D                                      10.15      4800                  22.50
C                                       8.24      5848                  27.42
B                                       6.82      2491                  11.68
Unreal                                  5.52      4092                  19.18
A                                       5.00      1501                   7.04

**Вывод:** 

Согласно проведенным рассчетам, между трудовым стажем и возвратом кредита в срок существует зависмость.

В зависмости от трудового стажа вероятность задолженности составляет от 5.00 % (стаж более 5000) до 11.24 % (стаж менее 500).

При этом отмечается прямая зависимость - с увеичением трудового стажа вероятность невозврата кредита в срок снижается.

Кроме того, следует обратить внимание на аномально большие значения столбца days_employed (4092 строки имеют значения days_employed более 10000 - каиегория "Ureal"). Вероятно некорректные значения получены в результате технической ошибки при импорте данных. 

<div class="alert alert-success">
<font size="5"><b>Комментарий ревьюера</b></font>

Успех: Круто, что ты проанализировал и этот пункт. 

</div>

#### 3.6 Есть ли зависимость между возрастом и возвратом кредита в срок

In [39]:
probability_of_debt_ver2('dob_years_category')

debt probability (%)  quantity  relative quantity (%)
dob_years_category                                                       
B                                  10.69      2123                   9.95
C                                   8.94     12494                  58.57
A                                   8.00       100                   0.47
D                                   5.75      6614                  31.01

Проверим корректность значений столбца dob_years

In [40]:
data['dob_years'].unique()

array([42, 36, 33, 32, 53, 27, 43, 50, 35, 41, 40, 65, 54, 56, 26, 48, 24,
       21, 57, 67, 28, 63, 62, 47, 34, 68, 25, 31, 30, 20, 49, 37, 45, 61,
       64, 44, 52, 46, 23, 38, 39, 51,  0, 59, 29, 60, 55, 58, 71, 22, 73,
       66, 69, 19, 72, 70, 74, 75])

In [41]:
data[data['dob_years']==0]['dob_years'].count()

100

**Вывод:** 

Согласно проведенным рассчетам, между возрастом и возвратом кредита в срок существует зависмость.

В зависмости от возраста вероятность задолженности составляет от 5.75 % до 10.69 %.

Наибольшая вероятность невозврата кредита в срок в категории "В" (19-27 лет).

Кроме того, 100 строк имеют значение dob_years == 0, что не может соответствовать действительности (возраст=0). Таким образом, в категорию "А" попали только некорректные данные.

Таким образом, можно сделать вывод о том, что с увеличением возраста вероятность неуплаты кредита в срок снижается.

<div class="alert alert-success">
<font size="5"><b>Комментарий ревьюера</b></font>

Успех: Хорошо бы все же не просто описывать результат, но также приводить и возможные причины его получения. В остальном все здорово. 
</div>

#### 3.7 Есть ли зависимость между полом заемщика и возвратом кредита в срок

In [42]:
probability_of_debt_ver2('gender')

debt probability (%)  quantity  relative quantity (%)
gender                                                       
M                      10.27      7238                  33.93
F                       7.02     14092                  66.06
XNA                     0.00         1                   0.00

**Вывод:** 

Согласно проведенным рассчетам, между полом заемщика и возвратом кредита в срок существует зависмость.

Вероятность задолженности для мужчин составляет 10.27 % против 7.02 % у женщин.

Отмечается одна строка с неивестной категорией "XNA".

#### 3.8 Есть ли зависимость между типом занятости и возвратом кредита в срок

In [43]:
probability_of_debt_ver2('income_type')

debt probability (%)  quantity  relative quantity (%)
income_type                                                           
в декрете                      100.00         1                   0.00
безработный                     50.00         2                   0.01
сотрудник                        9.57     11015                  51.64
компаньон                        7.41      5047                  23.66
госслужащий                      5.93      1451                   6.80
пенсионер                        5.67      3812                  17.87
предприниматель                  0.00         2                   0.01
студент                          0.00         1                   0.00

**Вывод:** 

Согласно проведенным рассчетам, между типом занятости и возвратом кредита в срок существует зависмость.

В зависмости от типа занятости вероятность задолженности составляет от 5.67 % (пенсионер) до 9.57 % (сотрудник).

При этом по ряду категорий (безоаботные, в декрете, предприниматель, студент) отсутствует достаточное количество данных, однако, вероятно, имеют высокие риски невозврата кредита в срок.

#### 3.9 Есть ли зависимость между уровнем образования и возвратом кредита в срок

In [44]:
probability_of_debt_ver2('education')

debt probability (%)  quantity  relative quantity (%)
education                                                                 
начальное                           10.99       282                   1.32
неоконченное высшее                  9.18       741                   3.47
среднее                              8.99     15075                  70.67
высшее                               5.32      5227                  24.50
ученая степень                       0.00         6                   0.03

**Вывод:** 

Согласно проведенным рассчетам, между уровнем образования и возвратом кредита в срок существует зависмость.

В зависмости от уровня образования вероятность задолженности составляет от 5.32 % (высшее) до 10.99 % (начальное).

При этом по категори "ученая степень" отсутствует достаточное количество данных.

<div class="alert alert-success">
<font size="5"><b>Комментарий ревьюера</b></font>

Успех: Круто, что ты проанализировал все характеристики клиентов 👍 

</div>

#### 3.5 Приведите возможные причины появления пропусков в исходных данных.

*Ответ:* Исходные данные содержат пропуски в столбцах 'days_employed' и 'total_income'. Причем, строки пропусков совпадают. На основании чего можно сделать вывод о том, что пропуски связаны с непредставлением сведений с места работы заемщика или утратой указанных сведений в результате технической ошибки при импорте данных.

<div class="alert alert-warning">
<font size="5"><b>Комментарий ревьюера</b></font>
  
Совет: Все же доход является важным пунктом при рассмотрении заявки. Больше похоже на то, что мы имеем техническую ошибку при импорте данных. Однако точного ответа тут нет, поэтому оставим текущий вариант. 

</div>

#### 3.6 Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.

*Ответ:* Медиана лучше всего отражает выборку количественных переменных в целом, так как на нее не влияют случайные выбросы и шум.

<div class="alert alert-success">
<font size="5"><b>Комментарий ревьюера</b></font>

Успех: Все верно, медиана более устойчива к наличию выбросов в данных.

</div>

### Шаг 4: общий вывод.

**Общий вывод.**



На основании проведенных рассчетов можно сделать вывод о том, что все представленные метрики в той или иной степени вляют на вероятность оплаты кредита в срок.

**Влияние установленных заданием метрик:**

- **Количество детей**

В зависмости от количества детей вероятность задолженности составляет от 7.54 % до 9.76 %. При этом не целесообразно учитывать вероятность 0 % для категории с 5 детьми ввиду малого количества сведений по указанной категории в представленной выборке (всего 9 займов).

Наименьшая вероятность задолженности у категории заемщиков без детей (7.54 %). При этом, заемщики указанной категории составляют большую часть представленной выборки (66.06 %).

Наибольшая вероятность отмечается в категории с 4 детьми (9.76 %).

При этом сделать вывод о том, что вероятность задолженности увеличивается с количеством детей не представляется возможным.

- **Семейное положение** 

В зависмости от семейного положения вероятность задолженности составляет от 6.62 % (вдовец / вдова) до 9.76 % (Не женат / не замужем).

При этом наиболее многочисленна категория "женат / замужем" (57.48 % выборки) с верочятностью задолженности 7.56 %.

Кроме того отмечается, что наименования категорий сатуса семейного положения могут вводить в заблуждение. Например, "в разводе" и "вдовец/вдова" могут принадлежать к категории "Не женат / не замужем". И наоборот. Что может повлиять на результаты исследования.

- **Уровень дохода** 

В зависмости от уровня дохода вероятность задолженности составляет от 6.02 % (кат. "D") до 8.50 % (кат."С"). При этом не целесообразно учитывать вероятность для категорий дохода "А" (8.00 %) и "Е" (9.09 %) ввиду малого количества сведений по указанным категориям в представленной выборке (всего 25 и 22 займа соответственно).

Наибольшая вероятность отмечается у заемщиков с категорией дохода "С" (8.50 %). При этом, заемщики указанной категории составляют большую часть представленной выборки (74.64 %).

Сделать вывод о том, что вероятность задолженности уменьшается с увеличением уровня дохода не представляется возможным.

- **Цель кредита** 

В зависмости от цели кредита вероятность задолженности составляет от 7.26 % (операции с недвижимостью) до 9.35 % (операции с автомобилем).

При этом наиболее многочисленна категория "операции с недвижимостью" (50.40 % выборки).

**Влияние дополнительных метрик:**

- *Трудовой стаж*

В зависмости от трудового стажа вероятность задолженности составляет от 5.00 % (стаж более 5000) до 11.24 % (стаж менее 500).

При этом отмечается прямая зависимость - с увеичением трудового стажа вероятность невозврата кредита в срок снижается.

Кроме того, следует обратить внимание на аномально большие значения столбца days_employed (4092 строки имеют значения days_employed более 10000 - каиегория "Ureal"). Вероятно некорректные значения получены в результате технической ошибки при импорте данных. 

- *Возраст*

В зависмости от возраста вероятность задолженности составляет от 5.75 % до 10.69 %.

Наибольшая вероятность невозврата кредита в срок в категории "В" (19-27 лет).

Кроме того, 100 строк имеют значение dob_years == 0, что не может соответствовать действительности (возраст=0). Таким образом, в категорию "А" попали только некорректные данные.

Таким образом, можно сделать вывод о том, что с увеличением возраста вероятность неуплаты кредита в срок снижается.

- *Пол заемщика*

Вероятность задолженности для мужчин составляет 10.27 % против 7.02 % у женщин.

Отмечается одна строка с неивестной категорией "XNA".

- *Тип занятости*

В зависмости от типа занятости вероятность задолженности составляет от 5.67 % (пенсионер) до 9.57 % (сотрудник).

При этом по ряду категорий (безоаботные, в декрете, предприниматель, студент) отсутствует достаточное количество данных, однако, вероятно, имеют высокие риски невозврата кредита в срок.

- *Уровень образования*

В зависмости от уровня образования вероятность задолженности составляет от 5.32 % (высшее) до 10.99 % (начальное).

При этом по категори "ученая степень" отсутствует достаточное количество данных.

<div class="alert alert-danger">
<font size="5"><b>Комментарий ревьюера</b></font>

Ошибка: Финальный вывод и есть главный результат твоей работы. Стоит писать его подробно по результатам проведенной работы. В задании у нас перечислены основные параметры, которые мы должны были проанализировать. Поэтому им в выводе стоит уделить больше внимания. Также можно расписать все, что было сделано в работе (этот пункт опциональный).

</div>

<div class="alert alert-success">
<font size="5"><b>Комментарий ревьюера 2</b></font>

Успех: Выводы описаны грамотно. Приведены ответы на главные вопросы проекта. В выводах можно приводить полученные ранее значения, правильно. Также можно расписать все, что было сделано в ходе проведения работы. Еще лучше будет, если приведешь рекомендации для компании по дальнейшим действиям. 

</div>

<div class="alert alert-info">
<font size="5"><b>Комментарий ревьюера</b></font>

Ты проделал большую работу, молодец! Ты овладел важными на данном этапе обучения инструментами, а также довольно качественно их используешь. Критических замечаний немного. Однако их важно исправить. Также есть несколько желтых комментариев. Думаю, исправить и их тебе не составит труда. Жду твою работу :)

</div>

<div class="alert alert-success">
<font size="5"><b>Комментарий ревьюера 2</b></font>

Теперь все сделано здорово. Ты проделал классную работу, молодец. Поздравляю с ее завершением. Надеюсь, проект был интересен и поставлен. Продолжай в том же духе. Успехов в дальнейшем пути :)
</div>